[Fix #8859] Add new `Lint/UnmodifiedReduceAccumulator` cop

dvandersluis opened the pull request:
Adds a cop to check for `reduce` or `inject` blocks that do not modify the accumulator. Closes #8859.

I tried to do this in the safest way, and avoid false positives as much as possible. The cop takes a look at all the return values from the block and makes some judgement on them (in order):
* If the block returns the accumulator indexed (ie. `reduce { |acc, el| acc[el] += 1 }`), it'll always register an offense because that will likely cause exceptions to be thrown.
* If the block returns the accumulator, or an expression containing it, in any branch, it is considered valid and no offenses will be registered.
* If the block returns an expression containing the element and another value (or just other values), it is considered valid because we cannot determine for sure what that value is.
* Otherwise, returns containing the element are registered as offenses. 

I think I covered a lot of edge cases (running this on rubocop itself helped there), but please let me know if I've missed something, I'd like to try to avoid having a new cop immediately have a bunch of bug reports 😅 

PS: I'm not 100% sold on the name and went back and forth for a while. Any suggests are appreciated.

-----------------

Before submitting the PR make sure the following are checked:

* [x] Wrote [good commit messages][1].
* [x] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Added an entry to the [Changelog](https://github.com/rubocop-hq/rubocop/blob/master/CHANGELOG.md) if the new code introduces user-observable changes. See [changelog entry format](https://github.com/rubocop-hq/rubocop/blob/master/CONTRIBUTING.md#changelog-entry-format).
* [x] The PR relates to *only* one subject with a clear title and description in grammatically correct, complete sentences.
* [x] Run `bundle exec rake default`. It executes all tests and RuboCop for itself, and generates the documentation.

[1]: https://chris.beams.io/posts/git-commit/

dvandersluis commented on config/default.yml:
```diff
@@ -1735,6 +1735,11 @@ Lint/RandOne:
+Lint/ReduceAccumulator:
+  Description: Checks for `reduce` or `inject` blocks that do not update the accumulator each iteration.
+  Enabled: pending
+  VersionAdded: '1.1'
```
@bbatsov not sure if this is the correct version now.

bbatsov replied:
It is.

dvandersluis commented on lib/rubocop/cop/lint/reduce_accumulator.rb:
```diff
@@ -0,0 +1,148 @@
+          }
+        PATTERN
+
+        def_node_search :expression_values, <<~PATTERN
+          `{
+            (%RuboCop::AST::Node::VARIABLES $_)
+            (%RuboCop::AST::Node::EQUALS_ASSIGNMENTS $_ ...)
+            (send (%RuboCop::AST::Node::VARIABLES $_) :<< ...)
+            (send nil? $_)
+            (dstr (begin {(send nil? $_) (%RuboCop::AST::Node::VARIABLES $_)}))
+            (%RuboCop::AST::Node::SHORTHAND_ASSIGNMENTS (%RuboCop::AST::Node::EQUALS_ASSIGNMENTS $_) ...)
+          }
+        PATTERN
```
@marcandre I feel like you probably have a simpler way to do this :grin:

bbatsov commented on lib/rubocop/cop/lint/reduce_accumulator.rb:
```diff
@@ -0,0 +1,148 @@
+      #     acc
+      #   end
+      class ReduceAccumulator < Base
+        MSG = 'Ensure the accumulator will be modified by `%<method>s`.'
```
Might be a good idea to mention the accumulator name explicitly in the messages.

bbatsov commented:
I like the cop, but I find the name confusing. Perhaps something like `UnusedReduceAccumulator`? I'm totally open to better suggestions myself. :D

marcandre commented on lib/rubocop/cop/lint/reduce_accumulator.rb:
```diff
@@ -0,0 +1,148 @@
+          node.arguments[index].node_parts[0]
+        end
+
+        # Look for the an index of the accumulator being returned
```
^an^

marcandre commented on lib/rubocop/cop/lint/reduce_accumulator.rb:
```diff
@@ -0,0 +1,148 @@
+        end
+
+        # Look for the an index of the accumulator being returned
+        # This is always an offense, in order to try to catch
```
unfinished sentence

marcandre commented on lib/rubocop/cop/lint/reduce_accumulator.rb:
```diff
@@ -0,0 +1,148 @@
+      # could be rewritten as such without a loop.
+      #
+      # Also catches instances where an index of the accumulator is returned, as
+      # this changes the type of object being retained.
```
this *may* change

marcandre commented on lib/rubocop/cop/lint/reduce_accumulator.rb:
```diff
@@ -0,0 +1,148 @@
+      # this changes the type of object being retained.
+      #
+      # NOTE: For the purpose of reducing false positives, this cop only flags
+      # `reduce` blocks where an expression
```
unfinished sentence

marcandre reviewed (changes requested):
I started a review, but I think this cop is too ambitious and it's in general not possible to do what it wants to do.
I fear that the risk of false positive is too high. Unless I'm mistaken, this would be a false positive, and it would be very difficult to distinguish:
```ruby
enum.inject do |acc, elem|
  x = [*acc, elem]
  x << 42 if foo
  x
end
```

dvandersluis commented:
@marcandre thanks for the feedback, your example actually won't trip the cop (I can add it as a test if you'd like, I have a whole section of `does not register an offense if the return value cannot be determined` tests already [here](https://github.com/rubocop-hq/rubocop/pull/8916/files#diff-0c9784fbdc0a5d6631a93ed08c611e5d5a570839ad04c4a2966ff02d344fc634R231)).

I missed it in the description, but the cop only flags return values where the element is the *only* variable, for this exact reason. I didn't want the cop to make any assumptions here so if it's not clearly just returning the element, it does not register.

dvandersluis commented:
> I like the cop, but I find the name confusing. Perhaps something like `UnusedReduceAccumulator`? I'm totally open to better suggestions myself. :D

@bbatsov I agree it's not the best name. `UnmodifiedReduceAccumulator`?

dvandersluis commented:
@bbatsov @marcandre I updated for your feedback so far. I've also renamed the cop to `UnmodifiedReduceAccumulator` for now, still happy to rename further if a better name comes up. I've also added more examples to the documentation.

bbatsov commented:
I like the new name and I'm fine with merging the cop in its current state, unless @marcandre has some objections.

marcandre commented on lib/rubocop/cop/lint/reduce_accumulator.rb:
```diff
@@ -0,0 +1,164 @@
+        PATTERN
+
+        def_node_search :expression_values, <<~PATTERN
+          `{
```
I imagine this is an error, as you are doing `def_node_search` which will already go through descendants, right?

dvandersluis replied:
Yep, my mistake!

marcandre commented on lib/rubocop/cop/lint/reduce_accumulator.rb:
```diff
@@ -0,0 +1,164 @@
+            (%RuboCop::AST::Node::VARIABLES $_)
+            (%RuboCop::AST::Node::EQUALS_ASSIGNMENTS $_ ...)
+            (send (%RuboCop::AST::Node::VARIABLES $_) :<< ...)
+            (send nil? $_)
```
This mistakes the name of a method with the name of a variable

dvandersluis replied:
This is actually needed, because I'm looking for identifiers in general, not specifically variables. This is valid code and should not be flagged since we can't determine what `method` is from static analysis:

```ruby
(1..4).reduce(0) do |acc, el|
  method + el
end
```

marcandre replied:
I'm confused as to why there's `nil?` then. For example `self.method + el`, or `el.method`?

marcandre replied:
(BTW, please do not "resolve conversation" when there's no agreement)

dvandersluis replied:
You're right there shouldn't be!

dvandersluis replied:
> (BTW, please do not "resolve conversation" when there's no agreement)

Sorry about that!

marcandre replied:
> You're right there shouldn't be!

Ok, if you remove `nil?`, then `el` and `self.el` are treated the same way, even though they are not related?

dvandersluis replied:
Good call, I've fixed that now too. (btw I've been taking all your notes and applying them locally for the purpose of a new fix PR)

marcandre commented on lib/rubocop/cop/lint/reduce_accumulator.rb:
```diff
@@ -0,0 +1,164 @@
+            (%RuboCop::AST::Node::EQUALS_ASSIGNMENTS $_ ...)
+            (send (%RuboCop::AST::Node::VARIABLES $_) :<< ...)
+            (send nil? $_)
+            (dstr (begin {(send nil? $_) (%RuboCop::AST::Node::VARIABLES $_)}))
```
Same here (first part of `{}`)

marcandre reviewed (commented):
Sorry I didn't have the time to reply earlier, and I do have objections.

It checks for two scenarios, both of which can yield false positives for perfectly valid code (in the review)

I think this should either be reverted (at least until it is revised, but again I think it is not possible to do so), or this cop should remain disabled by default.

dvandersluis commented on lib/rubocop/cop/lint/reduce_accumulator.rb:
```diff
@@ -0,0 +1,164 @@
+      # @example
+      #
+      #   # bad
+      #   (1..4).reduce(0) do |acc, el|
+      #     el * 2
+      #   end
```
>Bad example, acc is not used at all. There are already cops that deal with this, and Ruby too.

Do you mean `Lint/UnusedBlockArgument` here? While that's true it doesn't really convey the same thing that this cop does. (I don't see any other offenses for the example, so I think it's valid).

I think you meant false *positives* (false negatives are inherent here on purpose), but you're right that there is a class of false positives here that I did not consider. I'm looking at this now, thanks!

marcandre replied:
> I don't see any other offenses for the example, so I think it's valid

specs only check for the current cop, you can't rely on that. 

> I think you meant false positives 

Yes, sorry 😅

dvandersluis replied:
@marcandre no I meant I ran all of rubocop on the example haha, it raise `Lint/UnusedBlockArgument` and this cop. :grin:

dvandersluis commented on lib/rubocop/cop/lint/reduce_accumulator.rb:
```diff
@@ -0,0 +1,164 @@
+      #   end
+      #
+      #   # bad, may raise a NoMethodError after the first iteration
+      #   %w(a b c).reduce({}) do |acc, letter|
+      #     acc[letter] = true
+      #   end
```
I concede that that is a valid use that will be registered by the cop, but I'd argue that this is going to be *much* less prevalent than mistakenly forgetting to return the accumulator when setting a hash key.

I see a few options:
1. We can remove this behaviour (I think this isn't ideal, because I do think this is still a good check)
2. We can put it behind a conf variable that is not enabled by default
3. We can accept the false positives (if this was my code I'd just `# rubocop:disable` it, but maybe that's suboptimal).

What do we think? /cc @bbatsov

marcandre replied:
> forgetting to return the accumulator when setting a hash key.

BTW this is a double error, as `inject` probably shouldn't be used in the first place in that case (see `Style/EachWithObject`)

dvandersluis replied:
Definitely true, but in this case `Style/EachWithObject` doesn't register (because the accumulator is not being returned).

dvandersluis commented:
@marcandre I opened #8949 to continue this.
